database/sql/src/main/scala/cromwell/database/slick/WorkflowStoreSlickDatabase.scala

```diff
@@ -1,55 +1,49 @@
 
-object WorkflowStoreSlickDatabase {
-  case class NotInOnHoldStateException(workflowId: String) extends Exception(s"Couldn't change status of workflow $workflowId to 'Submitted' because the workflow is not in 'On Hold' state")
-}
-
 trait WorkflowStoreSlickDatabase extends WorkflowStoreSqlDatabase {
   this: EngineSlickDatabase =>
 
   import dataAccess.driver.api._
 
-  override def setAllRunningToAborting()(implicit ec: ExecutionContext): Future[Unit] = {
+  override def setStateToState(fromWorkflowState: String, toWorkflowState: String)
+                              (implicit ec: ExecutionContext): Future[Unit] = {
     val action = dataAccess
-      .workflowStateForWorkflowState(WorkflowStoreState.Running)
-      .update(WorkflowStoreState.Aborting)
-
-    runTransaction(action).void
-  }
-
-  override def markRunningAndAbortingAsRestarted(cromwellId: Option[String])(implicit ec: ExecutionContext): Future[Unit] = {
-    val action = dataAccess.heartbeatTimestamp(cromwellId).update(None)
+      .workflowStateForWorkflowState(fromWorkflowState)
+      .update(toWorkflowState)
 
     runTransaction(action).void
   }
 
   /**
     * Set the workflow Id to Aborting state.
-    * @return Some(restarted) if the workflow exists in the store, where restarted is true if the workflow's heartbeat
-    *         timestamp was cleared on restart.
-    *         None if the workflow does not exist in the store
+    *
+    * @param workflowExecutionUuid  Id to update or delete.
+    * @param workflowStateToDelete  Delete rows with this state.
+    * @param workflowStateForUpdate Update other rows to this state.
+    * @return Number of rows deleted, plus Some(heartbeatTimestampIsEmpty) if the workflow exists in the store, where
+    *         heartbeatTimestampIsEmpty is true if the workflow's heartbeat timestamp is empty, and None if the workflow
+    *         does not exist.
     */
-  override def setToAborting(workflowId: String)
-                            (implicit ec: ExecutionContext): Future[Option[Boolean]] = {
+  override def deleteOrUpdateWorkflowToState(workflowExecutionUuid: String,
+                                             workflowStateToDelete: String,
+                                             workflowStateForUpdate: String)
+                                            (implicit ec: ExecutionContext): Future[(Int, Option[Boolean])] = {
     val action =  for {
-      restarted <- dataAccess.heartbeatClearedForWorkflowId(workflowId).result.headOption
-      _ <- dataAccess.workflowStateForId(workflowId).update(WorkflowStoreState.Aborting)
-    } yield restarted
+      restarted <- dataAccess.heartbeatClearedForWorkflowId(workflowExecutionUuid).result.headOption
+      deleted <- dataAccess
+        .workflowStateForWorkflowExecutionUUidAndWorkflowState((workflowExecutionUuid, workflowStateToDelete))
+        .delete
+      _ <- dataAccess.workflowStateForWorkflowExecutionUUid(workflowExecutionUuid).update(workflowStateForUpdate)
```

Comment:
ToL:
Given that a (non quantum) workflow can only be in one state at a time I think we could avoid doing the update if the delete returns `1`, which would save us a little in DB requests.
Also I'm wondering (and really don't know the answer) if we expect most aborts to be coming for on hold workflows or `submitted/running`. If the latter we might want to reverse those 2 requests and try the update first.
This also makes me think, why do we not delete the entry when the status is `Submitted` too ?

Reply:
Reversing the rows is possible, but would require slightly more complex SQL.

Right now the flow looks like:

1. Delete the selected wfid with status 'On Hold' or 'Submitted'
2. (if deleted == 0) Update the selected wfid to status 'Aborting'

Reversing these and getting the same results would require a bit more complexity. Doable, but I'm not convinced it's worth the complexity, yet.

Reply:
That's fine by me, I'm 👍 with the current order if it's not worth the hassle

Reply:
On this point I think that @Horneth 's intuition that aborting normal workflows will happen **far** more frequently than on hold, so if there's a real performance hit that'd be not so great. 

It's at least worth considering from that perspective, but OTOH this could be more akin to my old "reorder the routes by frequency" performance enhancement which was never really going to matter

Reply:
I wouldn't worry about any of this since the update is going to be removed for DB-driven abort. 🙂